text/3594-expose-stackrealign-attribute.md

```diff
@@ -0,0 +1,83 @@
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+The `[realign_stack]` attribute can be added to a function to tell the compiler to add stack re-alignment to that function if necessary.
+Useful in cases where your code is called from a thread or a binary compiled with another compiler, that uses different alignment and thus lead to a corruption.
+An example of one such setting could be `-mpreferred-stack-boundary=2` in GCC which would set the stack alignment to 4 instead of the default value for the ABI which is 16.
+Other such settings could be present at GCC's "Machine-Dependent Options", which there are many of, and many of them can break ABI compatibility.
+
+```
+#[realign_stack]
+#[no_mangle]
+pub extern "C" fn callback_function() -> i32 {
+    println!("Called from callback!!");
+
+    0
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+The realign_stack attribute is specified as follows:
+```
+#[realign_stack]
+```
+
+When the `realign_stack` attribute is applied to a function, the compiler no longer assumes the stack is properly aligned when the function is called, so will insert code to align the stack as needed for calling other functions, variables requiring alignment, etc.
+This alignment is achieved by adjusting the stack pointer accordingly. The alignment is maintained for the duration of the function's execution.
+Adding this attribute unnecessarily might "waste" space on the stack which could be crucial in real-time systems.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+- Introducing a new attribute adds complexity to the language.
+- Limited use cases: Stack realignment may not be necessary in most Rust codebases, potentially making this feature less relevant for many developers.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+An alternative could be a macro workaround instead of adding the attribute.
+However it would be more like band-aid than an actual solution.
+Another alternative could be adding the any extern "C" function the `stackrealign` attribute implicitly which would solve the main use-case.
+An extra option could be not verifying data-layout for custom targets provided via `--target=`, which would allow users to patch the "natural stack alignment" in their custom target which should relax LLVM stack alignment assumptions that are present in the system.
+Another alternative could be adding a new ABI that captures "function which can be called with any stack alignment".
+I chose to propose this RFC and not any of the alternatives because it seems to me that this proposition provides the simplest solution, a solution that is very close to `force_align_arg_pointer` function attribute in GCC and a solution that is easy to implement for rustc.

```

Comment:
"easy to implement" and "it's like what GCC does" are not very strong arguments, IMO. The question is what is best for Rust programmers.

@Lokathor raised some points for why they did not think a separate ABI was a good idea; those should be reflected here.

Reply:
Added more explaining there